Print when columns are ignored during CSV import

spacecowboy opened the pull request:
This change warns the user about a typical symptom of an incorrect header being used (for example using different delimiters in header versus data). After some thinking (and implementing a version with StringLogger) I came to the decision that a simple System.out.println is exactly what we want here. Doing almost anything else will introduce more changes elsewhere and be harder to test for.

Example of the output:

```
Importing the contents of these files into /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db:
Nodes:
  /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/0-nodes-header.csv
  /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/1-nodes.csv
Relationships:
  /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/2-relationships-header.csv
  /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/3-relationships.csv

Available memory:
  Free machine memory: 288.05 MB
  Max heap memory : 910.50 MB

Nodes

[>:??----------|PROPERTIES-|NODE:7.63 MB--|*LABEL SCAN---------------------------------------|v] 10k
Done in 340ms
Warning: Ignored 3 extra columns in /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/1-nodes.csv
Prepare node index

[*DETECT:7.63 MB-------------------------------------------------------------------------------]   0
Done in 24ms
Calculate dense nodes
[...]
```

spacecowboy committed:
Print when columns are ignored during CSV import

MishaDemianenko commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -920,12 +945,22 @@ private void writeNodeData( PrintStream writer, Configuration config, List<Strin
         }
     }
 
+    private String getLine(String nodeId, char delimiter, char arrayDelimiter, int columnCount) {
```
last param called columnCount while all others method in a call chain use `extraColumns` can we align those?

MishaDemianenko commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -920,12 +945,22 @@ private void writeNodeData( PrintStream writer, Configuration config, List<Strin
+                delimiter + randomLabels( arrayDelimiter ));
+
+        for ( int i = 0; i < columnCount; i++ ) {
+            stringBuilder.append( delimiter ).append( "ExtraCol").append( i );
```
Why not just say `ExtraColumn`? what the reason to same couple of chars?

tinwelint commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/InputEntityDeserializer.java:
```diff
@@ -176,6 +186,11 @@ private boolean deserializeNextFromSource() throws IOException
     {
+        if ( ignoredColumns > 0 )
+        {
+            System.out.println( String.format( "Warning: Ignored %d extra columns in %s", ignoredColumns,
```
Yeah, perhaps gather line numbers (at least the first 10 or something), and some sample data?

thobe replied:
I'd prefer to get a histogram output of the ignored columns.

Something like:

```
Warning: Ignored 2 extra columns in 4 lines of /your/file.csv
Warning: Ignored 5 extra columns in 3 lines of /your/file.csv
```

Maintaining that state is easy enough, and it helps immensely in figuring out what the reason for the extra columns might be. Especially if we also print the total number of lines of the file.

spacecowboy replied:
Having tested both, I vote for @tinwelint 's suggestion. Seems to be the most useful. Example from the test with variable number of extra columns:

```
>:??--------------------------|PROPERTI|NODE:7.63 MB|*LABEL SCAN----------------------------|v] 10k
Done in 292ms
Warning: Ignored several extra columns in 
    /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/1-nodes.csv
If this is unexpected, check the headers. Example of ignored values:
Row 1:  ExtraColumn0    ExtraColumn1
Row 2:  ExtraColumn0    ExtraColumn1
Row 3:  ExtraColumn0
Row 4:  ExtraColumn0
Row 7:  ExtraColumn0
Row 8:  ExtraColumn0    ExtraColumn1
Row 10:     ExtraColumn0
Row 11:     ExtraColumn0
Row 13:     ExtraColumn0

Prepare node index

[*DETECT:7.63 MB-------------------------------------------------------------------------------]   0
```

thobe commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -151,6 +149,27 @@ public void shouldImportWithHeadersBeingInSeparateFiles() throws Exception
+                        .getAbsolutePath() + MULTI_FILE_DELIMITER + nodeData( false, config, nodeIds, alwaysTrue(),
+                        Charset.defaultCharset(), extraColumns ).getAbsolutePath(), "--relationships",
+                relationshipHeader( config ).getAbsolutePath() + MULTI_FILE_DELIMITER + relationshipData( false, config,
+                        nodeIds, alwaysTrue(), true ).getAbsolutePath() );
```
This is too dense to read. Let's put this one option per line.

Something like:

```
executeImportAndCatchOutput(
        "--into", dbRule.getStoreDirAbsolutePath(),
        "--delimiter", "TAB",
        "--array-delimiter", String.valueOf( config.arrayDelimiter() ),
        "--nodes", nodeHeader( config ).getAbsolutePath() + MULTI_FILE_DELIMITER
                + nodeData( false, config, nodeIds, alwaysTrue(), Charset.defaultCharset(), extraColumns ).getAbsolutePath(),
        "--relationships", relationshipHeader( config ).getAbsolutePath() + MULTI_FILE_DELIMITER
                + relationshipData( false, config, nodeIds, alwaysTrue(), true ).getAbsolutePath() );
```

(but perhaps more line breaks are needed)

thobe commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -920,12 +945,22 @@ private void writeNodeData( PrintStream writer, Configuration config, List<Strin
 
+    private String getLine(String nodeId, char delimiter, char arrayDelimiter, int extraColumns) {
+        StringBuilder stringBuilder = new StringBuilder(  nodeId  + delimiter + randomName() +
+                delimiter + randomLabels( arrayDelimiter ));
```
Don't do string concatenation along with a StringBuilder, this will compile down to using two StringBuilders, which is completely unnecessary.

Sure, this is "only a test", but we should hold those up to the same standard as the rest of the code. If for no other reason then at least so that we remember to always do it right.

Also: passing a string to the constructor of a StringBuilder typically isn't very useful since it'll set the buffer size of the StringBuilder to the size of the string, and you probably want the resulting string to be larger than that, so you want a larger buffer, otherwise you'll get more re-allocations of the buffer.
What is useful for passing into the constructor is a template string in which you will make substitutions.

thobe replied:
A shorter way of saying what I said above would be to say that you should instead have written:

```
StringBuilder stringBuilder = new StringBuilder().append( nodeId ).append( delimiter ).append( randomName() ).append( delimiter ).append( randomLabels( arrayDelimiter ) );
```

thobe commented:
Needs rebasing.

spacecowboy commented:
@thobe now rebased and updated. Example of current output:

```
Warning: ignored columns which were not present in header for
    /home/jonas/workspace/2.2/community/import-tool/target/test-data/org.neo4j.tooling.ImportToolTest/graph-db/1-nodes.csv
Row 1:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 2:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 3:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 4:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 5:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 6:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 7:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 8:  ExtraColumn0    ExtraColumn1    ExtraColumn2
Row 9:  ExtraColumn0    ExtraColumn1    ExtraColumn2
...
```

MishaDemianenko commented:
@spacecowboy @tinwelint would it make sense to create output result file in cases when we exceed certain amount of incorrect data?
Current behaviour will not be really helpful for huge file import problems.

tinwelint commented:
@MishaDemianenko yeah, there's certainly a point to that. We could log these lines into the `bad` file, no?

tinwelint commented:
@spacecowboy wdyt about logging into `bad` file instead?

MishaDemianenko commented:
@spacecowboy @tinwelint sounds good for me

spacecowboy commented:
@tinwelint @MishaDemianenko we could absolutely log more details in `bad`, but it will turn a small change into a large one. Will implement and get back to you

spacecowboy committed:
Handle extra columns in CSV files using bad file

Now either fails with an exception or logs to bad file.

tinwelint commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java:
```diff
@@ -128,12 +128,11 @@ public void doImport( Input input ) throws IOException
         CountingStoreUpdateMonitor storeUpdateMonitor = new CountingStoreUpdateMonitor();
         try ( BatchingNeoStore neoStore = new BatchingNeoStore( fileSystem, storeDir, config,
                 writeMonitor, logging, monitors, writerFactory, additionalInitialIds );
-              OutputStream badOutput = new BufferedOutputStream( fileSystem.openAsOutputStream( badFile, false ) );
```
Looks like you didn't clean up unused imports. Do you have such settings in your IDE enabled?

tinwelint commented:
Looks good

spacecowboy committed:
Remove unused imports

spacecowboy commented:
@tinwelint removed those unused imports.
